/* Protocol between libxsocket and Xcache */

message xcache_cmd {
	enum req_type {
		XCACHE_FLAG_NOTIFSOCK = 1;
		XCACHE_FLAG_DATASOCKET = 2;
		XCACHE_ALLOC_CONTEXT = 3;
		XCACHE_STORE = 4;
		XCACHE_NEWSLICE = 5;
		XCACHE_FETCHCHUNK = 7;
		XCACHE_TIMEOUT = 8;
		XCACHE_CLEAR = 9;
		XCACHE_RESPONSE = 10;
		XCACHE_ERROR = 11;
		XCACHE_GET_STATUS = 12;
		XCACHE_READ = 13;
		XCACHE_SENDCHUNK = 14;
		XCACHE_FREE_CONTEXT = 15;
		XCACHE_EVICT = 16;
		XCACHE_CACHE = 17;
		XCACHE_STORE_NAMED = 18;
		XCACHE_FETCHNAMEDCHUNK = 19;
		XCACHE_PUSHCHUNK = 20;
		XCACHE_NEWPROXY = 21;
		XCACHE_ISLOCAL = 22;
		XCACHE_ENDPROXY = 23;
	};

	enum status_type {
		XCACHE_OK = 1;
		XCACHE_ERR_EXISTS = 2;
		XCACHE_INVALID_CID = 3;
		XCACHE_CID_NOT_FOUND = 4;
		XCACHE_CID_IN_USE = 5;
		XCACHE_CID_MARKED_FOR_DELETE = 6;
		XCACHE_NCID_CALC_ERR = 7;
	};

	required req_type cmd = 1;

	optional int32 context_id = 2;
	optional string cid = 3;
	optional bytes dag = 4;
	optional bytes data = 5;
	optional int32 ttl = 6;
	optional int32 cache_size = 7;
	optional int32 cache_policy = 8;
	optional int32 flags = 9;
	optional status_type status = 10;
	optional int32 readoffset = 11;
	optional int32 readlen = 12;
	optional bytes msg = 13;
	optional bytes publisher_name = 16;
	optional bytes content_name = 17;
	optional int32 proxy_id = 18;
	optional bytes source_dag = 19;
};

message xcache_notif {
	required int32 cmd = 1;
	oneof notif_type{
		notif_arrived arrived = 2;
		notif_contents contents = 3;
	};
};

message notif_arrived {
	required bytes dag = 1;
};

message notif_contents {
	required string cid = 1;
	required bytes data = 2;
};
